### PR TITLE
Labels: reduce memory by de-duplicating strings in multiple SymbolTables

### PR DESCRIPTION
**Pull request by bboreham**
This builds on #10991, implementing "Option 5" in the [design doc](https://docs.google.com/document/d/1fXrPI2kr-jUvkF9JkkA7aPk7UQYTG9Yvk6A-fSinqEU/edit?usp=sharing).

The code adds a new `labels.Labels` implementation which can be selected at build time with `-tags dedupelabels`.

Previous attempts to do string interning in Prometheus have used a single lookup table for the entire program; this PR allows different sets of labels to have their own tables. Example: where you fetch series via remote-read, or receive series via remote-write with a symbol-table (see #11999).

Most of the code changes outside of the `model/labels` package are to manage the symbol tables.

Managing concurrent access to symbol tables:
* When creating new labels, a `sync.Mutex` is locked around reads and writes.
* When reading labels, there is no locking because the table of strings used by those labels is immutable.

~This branch includes two other PRs #13452 and #13491.~ Update: now merged.

### CODE REVIEW

**Comment by colega** [comment]
Can you explain in a few words how this deals with the churn of series in the Head?

File: model/histogram/float_histogram.go [hunk 1]
```diff
@@ -276,7 +276,7 @@ func (h *FloatHistogram) Sub(other *FloatHistogram) *FloatHistogram {
 // but they must represent the same bucket layout to match.
 func (h *FloatHistogram) Equals(h2 *FloatHistogram) bool {
 	if h2 == nil {
-		return false
+		return h == nil
```
Comment (colega) [inline model/histogram/float_histogram.go:279]: This looks unrelated to this PR.

**Review by bboreham — COMMENTED** [review]
Thanks for taking a look!

> Can you explain in a few words how this deals with the churn of series in the Head?

Not implemented yet.  Suggestion: after head compaction and series GC, create a new `SymbolTable` then run through all live series and rebuild their `Labels`.
Then reset the `SymbolTable`s used by scraping parsers.

File: model/labels/labels_string.go [hunk 2]
```diff
@@ -423,27 +527,22 @@ func FromStrings(ss ...string) Labels {
...
-	}
-
+	la, lb := len(a.data), len(b.data)
 	ia, ib := 0, 0
-	for ia < l {
+	for ia < la && ib < lb {
 		var aName, bName string
-		aName, ia = decodeString(a.data, ia)
-		bName, ib = decodeString(b.data, ib)
+		aName, ia = decodeString(a.syms, a.data, ia)
```
Comment (fpetkovski) [inline model/labels/labels_string.go:535]: If the two symbol tables are equal, would it be possible to avoid decoding label symbols and compare them directly?

Comment (bboreham) [reply model/labels/labels_string.go:535]: For `Equal`, yes, but not `Compare` since we don’t have an ordering in the symbol table. 
Might be worth skipping all the strings that are equal and just decoding the first one that’s different.

Comment (fpetkovski) [reply model/labels/labels_string.go:535]: That's a good point, decoding only what is different would probably speed things up when doing a merge of two sorted label sets. For example, when merging data from two TSDBs.

Comment (bboreham) [reply model/labels/labels_string.go:535]: Similar idea for current main: #12451

**Comment by bboreham** [comment]
I have rebased the code, and made some important updates since the last review:
* Prometheus can be built with either `-tags stringlabels` or `-tags internlabels` for the version in this PR.
* The symbol-table used for scraping is cleared out periodically, after a target reload.  (TODO: clear it out if targets are never reloaded)
* Symbol table locking was removed on the read path.

File: model/labels/labels_internlabels.go [hunk 3]
```diff
@@ -0,0 +1,778 @@
...
+	if i, found := t.byName[name]; found {
+		return i
+	}
+	i := t.nextNum
+	if t.nextNum == cap(t.byNum) {
+		// Name table is full; copy to a new one.  Don't touch the existing slice.
+		oldSlice := t.byNum
+		t.nameTable = &nameTable{byNum: make([]string, cap(oldSlice)*2), symbolTable: t}
+		copy(t.nameTable.byNum, oldSlice)
+	}
```
Comment (colega) [inline model/labels/labels_internlabels.go:84]: Can you explain here why we don't want to touch the existing slice, and why a simple append wouldn't work?

File: model/labels/labels_internlabels.go [hunk 4]
```diff
@@ -0,0 +1,778 @@
...
+	name = strings.Clone(name)
+	t.byNum[i] = name
+	t.byName[name] = i
+	t.nextNum++
+	return i
+}
+
+// ToName maps an integer to a string.
+func (t *nameTable) ToName(num int) string {
```
Comment (colega) [inline model/labels/labels_internlabels.go:93]: I feel like this should be `ToValue`, but it's opinionated.

File: model/labels/labels_internlabels.go [hunk 5]
```diff
@@ -0,0 +1,778 @@
...
+		// Just panic if we go of the end of data, since all Labels strings are constructed internally and
+		// malformed data indicates a bug, or memory corruption.
+		b := data[index]
+		index++
+		size |= int(b&0x7F) << shift
+		if b < 0x80 {
+			break
+		}
+	}
+	return size, index
```
Comment (colega) [inline model/labels/labels_internlabels.go:115]: Nit, I think this should be more generic:
```suggestion
	value := int(b & 0x7F)
	for shift := uint(7); ; shift += 7 {
		// Just panic if we go of the end of data, since all Labels strings are constructed internally and
		// malformed data indicates a bug, or memory corruption.
		b := data[index]
		index++
		value |= int(b&0x7F) << shift
		if b < 0x80 {
			break
		}
	}
	return value, index
```

File: model/labels/labels_internlabels.go [hunk 6]
```diff
@@ -0,0 +1,778 @@
...
+		b = append(b, seps[0])
+		b = append(b, value...)
+		b = append(b, seps[0])
+	}
+	return xxhash.Sum64(b), b
+}
+
+// BytesWithLabels is just as Bytes(), but only for labels matching names.
+// 'names' have to be sorted in ascending order.
+func (ls Labels) BytesWithLabels(buf []byte, names ...string) []byte {
```
Comment (colega) [inline model/labels/labels_internlabels.go:245]: Sounds like this should be `BytesForLabels` for consistency with `HashForLabels`, but feel free to ignore me if this is an existing contract.

File: model/labels/labels_internlabels.go [hunk 7]
```diff
@@ -0,0 +1,778 @@
...
+		}
+		pos = newPos
+	}
+	return b.Bytes()
+}
+
+// Copy returns a copy of the labels.
+func (ls Labels) Copy() Labels {
+	buf := append([]byte{}, ls.data...)
+	return Labels{syms: ls.syms, data: yoloString(buf)}
```
Comment (colega) [inline model/labels/labels_internlabels.go:297]:
```suggestion
	return Labels{syms: ls.syms, data: strings.Clone(ls.data)}
```

File: model/labels/labels_internlabels.go [hunk 8]
```diff
@@ -0,0 +1,778 @@
...
+	if name == "" { // Avoid crash in loop if someone asks for "".
+		return "" // Prometheus does not store blank label names.
+	}
+	for i := 0; i < len(ls.data); {
+		var lName, lValue string
+		lName, i = decodeString(ls.syms, ls.data, i)
+		if lName == name {
+			lValue, _ = decodeString(ls.syms, ls.data, i)
+			return lValue
+		} else if lName[0] > name[0] { // Stop looking if we've gone past.
```
Comment (colega) [inline model/labels/labels_internlabels.go:312]: Are we checking only `[0]` as a performance optimization? Does this play well with Unicode (seems like U nicode support is coming?)

Is it really faster than `lName > name`? (I guess CPUs have that optimized)

File: model/labels/labels_internlabels.go [hunk 9]
```diff
@@ -0,0 +1,778 @@
...
+		var lName string
+		lName, i = decodeString(ls.syms, ls.data, i)
+		if lName == name {
+			return true
+		} else if lName[0] > name[0] { // Stop looking if we've gone past.
+			break
+		}
+		_, i = decodeVarint(ls.data, i)
+	}
+	return false
```
Comment (colega) [inline model/labels/labels_internlabels.go:335]: Sounds like this entire method could be just this, with no performance impact:
```suggestion
	return ls.Get(name) != ""
```

File: model/labels/labels_internlabels.go [hunk 10]
```diff
@@ -0,0 +1,778 @@
...
+// WithoutEmpty returns the labelset without empty labels.
+// May return the same labelset.
+func (ls Labels) WithoutEmpty() Labels {
+	if ls.IsEmpty() {
+		return ls
+	}
+	// Idea: have a constant symbol for blank, then we don't have to look it up.
+	blank, ok := ls.syms.symbolTable.byName[""]
+	if !ok { // Symbol table has no entry for blank - none of the values can be blank.
+		return ls
```
Comment (colega) [inline model/labels/labels_internlabels.go:363]: I think that the "Idea:" would perform worse, as if we don't look it up, we have to check whether the constant symbol is present in `data`. I like this hack.

File: model/labels/labels_internlabels.go [hunk 11]
```diff
@@ -0,0 +1,778 @@
...
+		bName, ib = decodeString(b.syms, b.data, ib)
+		if aName != bName {
+			return false
+		}
+		var aValue, bValue string
+		aValue, ia = decodeString(a.syms, a.data, ia)
+		bValue, ib = decodeString(b.syms, b.data, ib)
+		if aValue != bValue {
+			return false
+		}
```
Comment (colega) [inline model/labels/labels_internlabels.go:412]: Do we really need this bit of code? We know `a.data` has pairs of strings, we can use just one string decoding in the loop.

File: model/labels/labels_internlabels.go (hunk 3 above)
Comment (bboreham) [reply model/labels/labels_internlabels.go:84]: So that we can read it (in `ToName`) without locking.

File: model/labels/labels_internlabels.go (hunk 6 above)
Comment (bboreham) [reply model/labels/labels_internlabels.go:245]: Yeah this is pre-existing for years, called from PromQL.

File: model/labels/labels_internlabels.go (hunk 3 above)
Comment (colega) [reply model/labels/labels_internlabels.go:84]: Oh I see, it's because `nameTable` is immutable. 

Let me post this very opinionated suggestion:

```suggestion
	if t.nextNum == cap(t.byNum) {
		// Name table is full; copy to a new one. Don't touch the existing slice as nameTable is immutable after construction.
		newSlice := make([]string, cap(t.byNum)*2)
		copy(newSlice, t.byNum)
		t.nameTable = &nameTable{byNum: newSlice, symbolTable: t}
	}
```

The reason behind this suggestion is that as a reader I want to see that we're creating a _new slice_, and setting a _new name table_. I find your current code kind of misleading from the point in the fact that it creates a reference to an oldSlice (are we going to do something with it?), then we set a new empty table and then we copy. As I said, very opinonated, but at least I'd state in the comment why are we doing this.

File: model/labels/labels_internlabels.go (hunk 8 above)
Comment (bboreham) [reply model/labels/labels_internlabels.go:312]: This one is significantly faster on the benchmark:

With `[0]`:
```
BenchmarkLabels_Get/with_5_labels/not-found_label/get-4         50844589                23.60 ns/op            0 B/op          0 allocs/op
BenchmarkLabels_Get/with_5_labels/not-found_label/has-4         49705240                23.91 ns/op            0 B/op          0 allocs/op
BenchmarkLabels_Get/with_10_labels/not-found_label/get-4        49250290                23.99 ns/op            0 B/op          0 allocs/op
BenchmarkLabels_Get/with_10_labels/not-found_label/has-4        50232136                23.74 ns/op            0 B/op          0 allocs/op
BenchmarkLabels_Get/with_30_labels/not-found_label/get-4        51744460                24.15 ns/op            0 B/op          0 allocs/op
BenchmarkLabels_Get/with_30_labels/not-found_label/has-4        49727462                24.52 ns/op            0 B/op          0 allocs/op
```

Without:
```
BenchmarkLabels_Get/with_5_labels/not-found_label/get-4         29063010                39.48 ns/op            0 B/op          0 allocs/op
BenchmarkLabels_Get/with_5_labels/not-found_label/has-4         30483260                45.85 ns/op            0 B/op          0 allocs/op
BenchmarkLabels_Get/with_10_labels/not-found_label/get-4        27639243                46.17 ns/op            0 B/op          0 allocs/op
BenchmarkLabels_Get/with_10_labels/not-found_label/has-4        30595272                46.34 ns/op            0 B/op          0 allocs/op
BenchmarkLabels_Get/with_30_labels/not-found_label/get-4        30075024                40.38 ns/op            0 B/op          0 allocs/op
BenchmarkLabels_Get/with_30_labels/not-found_label/has-4        29790415                39.70 ns/op            0 B/op          0 allocs/op
```

File: model/labels/labels_internlabels.go (hunk 9 above)
Comment (colega) [reply model/labels/labels_internlabels.go:342]: Interesting and surprising. The impact seems to be a quite consistent 4ns increase, right? I guess that's the cost of the extra `decodeString` call for getting the value.

File: model/labels/labels_internlabels.go (hunk 7 above)
Comment (bboreham) [reply model/labels/labels_internlabels.go:308]: Same change can be made in `labels_stringlabels.go`.

**Comment by bboreham** [comment]
Pretty good since the last change, which I excerpted as #13427.  (Sorting of series is slower when it has to indirect through the symbol-table, but prombench never actually needs to sort series in Select).

image 
image
image 
image

**Comment by bboreham** [comment]
@colega asked a long time ago, but at the time there was no answer:

> Can you explain in a few words how this deals with the churn of series in the Head?

Now, each scrape target has its own symbol table. If a target disappears, its series will be removed from the head after 2-3 hours and the table with all strings will be garbage-collected.

If a target continually exports new labels, its symbol table will grow.  We detect that in `scrapePool.reload()` and start a new table periodically.  This may need improving for targets that never get reloaded.  
If series from different ages of the target continue to exist, they will keep the symbol table from that time alive on the heap. If we see cases of that I can look at something else, maybe building a new symbol-table for all series in the head after compaction so all.

In Prombench, which does simulate quite a lot of churn in targets, no growth in symbol table memory was observed.

File: model/labels/labels_internlabels.go (hunk 4 above)
Comment (bboreham) [reply model/labels/labels_internlabels.go]: Given we have explicit names and values for labels, I don't think I should prefer one here.
I can't remember why it's a "name table" and not a "string table".

File: model/labels/labels_dedupelabels.go [hunk 12]
```diff
@@ -0,0 +1,807 @@
+// Copyright 2017 The Prometheus Authors
```
Comment (SuperQ) [inline model/labels/labels_dedupelabels.go:1]:
```suggestion
// Copyright 2024 The Prometheus Authors
```

**Comment by bboreham** [comment]
I forgot I need to hack in `-tags dedupelabels` before running Prombench.

**Comment by bboreham** [comment]
Memory is about 15% lower, as hoped.  CPU is less obvious; sometimes the new one is using more and sometimes the old one.
However the query stats say the new one executes range queries faster, so is doing more work.

image

**Comment by bboreham** [comment]
Thanks for looking at this.  I really need a bit more context for the profile, to see where those functions were called. Could you upload the raw data?

**Comment by prymitive** [comment]
Is an svg ok?

pprof001

**Comment by bboreham** [comment]
Svg is better but I still can't see the line numbers.  However in this case I think it's enough; the main growth is under the scrape loop.

**Comment by prymitive** [comment]
> Svg is better but I still can't see the line numbers. However in this case I think it's enough; the main growth is under the scrape loop.

I suspect it's the `scrapeCache` as it keeps references to labels.

```
ROUTINE ======================== github.com/prometheus/prometheus/scrape.(*scrapeCache).addDropped in scrape/scrape.go
    1.78GB     1.78GB (flat, cum)  2.94% of Total
         .          .    974:func (c *scrapeCache) addDropped(met []byte) {
      93MB       93MB    975:	iter := c.iter
    1.69GB     1.69GB    976:	c.droppedSeries[string(met)] = &iter
         .          .    977:}
         .          .    978:
         .          .    979:func (c *scrapeCache) getDropped(met []byte) bool {
         .          .    980:	iterp, ok := c.droppedSeries[string(met)]
         .          .    981:	if ok {
ROUTINE ======================== github.com/prometheus/prometheus/scrape.(*scrapeCache).addRef in scrape/scrape.go
    5.31GB     5.31GB (flat, cum)  8.78% of Total
         .          .    967:func (c *scrapeCache) addRef(met []byte, ref storage.SeriesRef, lset labels.Labels, hash uint64) {
         .          .    968:	if ref == 0 {
         .          .    969:		return
         .          .    970:	}
    5.31GB     5.31GB    971:	c.series[string(met)] = &cacheEntry{ref: ref, lastIter: c.iter, lset: lset, hash: hash}
         .          .    972:}
         .          .    973:
         .          .    974:func (c *scrapeCache) addDropped(met []byte) {
         .          .    975:	iter := c.iter
         .          .    976:	c.droppedSeries[string(met)] = &iter
ROUTINE ======================== github.com/prometheus/prometheus/scrape.(*scrapeCache).trackStaleness in scrape/scrape.go
    3.75GB     3.75GB (flat, cum)  6.20% of Total
         .          .    987:func (c *scrapeCache) trackStaleness(hash uint64, lset labels.Labels) {
    3.75GB     3.75GB    988:	c.seriesCur[hash] = lset
         .          .    989:}
         .          .    990:
         .          .    991:func (c *scrapeCache) forEachStale(f func(labels.Labels) bool) {
         .          .    992:	for h, lset := range c.seriesPrev {
         .          .    993:		if _, ok := c.seriesCur[h]; !ok {

```

Could it be possible that SymbolTable instances are being kept referenced there for too long (forever?) ?

**Comment by bboreham** [comment]
Yes that is a good point: `scrapeCache .. cacheEntry .. lset` will hold SymbolTables alive.

I could flush the cache at the point it decides to discard the old SymbolTable.
https://github.com/prometheus/prometheus/blob/9e2c335bab/scrape/scrape.go#L359

**Comment by prymitive** [comment]
I wonder if the unique package added to the next go release would help here - https://pkg.go.dev/unique@master

**Comment by bboreham** [comment]
That one gets you to 8 bytes per string, or 16 per label.
This PR is more like 4.5 per label.

However, it might be preferable considering the amount of work needed to get this one to behave.

**Comment by bboreham** [comment]
Small update:
All the fixes to continuous memory growth have landed, and I plan to push a a 2.54.0+dedupelabels image next week.
According to PromBench at #14462 the memory saving is now only around 10%, which I think is because `mem_series` got bigger.
I am experimenting with a couple of ideas to bring that back.

(If you have a lot of long and non-unique labels then your saving will look better)

**Comment by bboreham** [comment]
The new Docker image is `prom/prometheus:v2.54.0-dedupelabels` - did anyone else try it?

**Comment by SuperQ** [comment]
Not yet, I will give it a try soon

**Comment by prymitive** [comment]
> The new Docker image is `prom/prometheus:v2.54.0-dedupelabels` - did anyone else try it?

I have a pair of identical servers running 2.54.0 - one with stringlabels & the other with dedupelabels, I think I need to give it a week or two to tell anything conclusive since last time I tried it it was slowly using more memory over time - https://github.com/prometheus/prometheus/pull/12304#issuecomment-2024756897

image

So far the heap profile looks good, without SymbolTable getting to the top:

```
File: prometheus
Type: inuse_space
Time: Aug 23, 2024 at 9:12am (UTC)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top
Showing nodes accounting for 30.53GB, 61.02% of 50.02GB total
Dropped 742 nodes (cum <= 0.25GB)
Showing top 10 nodes out of 127
      flat  flat%   sum%        cum   cum%
    5.23GB 10.46% 10.46%     5.23GB 10.46%  github.com/prometheus/prometheus/scrape.(*scrapeCache).addRef
    4.99GB  9.97% 20.43%     5.91GB 11.82%  github.com/prometheus/prometheus/tsdb.newMemSeries
    3.80GB  7.60% 28.03%     3.80GB  7.60%  github.com/prometheus/prometheus/scrape.(*scrapeCache).trackStaleness
    3.52GB  7.04% 35.07%     3.52GB  7.04%  github.com/prometheus/prometheus/tsdb/chunkenc.NewXORChunk
    2.73GB  5.46% 40.53%     2.73GB  5.46%  github.com/prometheus/prometheus/tsdb/chunkenc.(*bstream).writeByte
    2.58GB  5.17% 45.69%     2.58GB  5.17%  github.com/prometheus/prometheus/model/labels.(*SymbolTable).toNumUnlocked
    2.14GB  4.27% 49.96%     2.14GB  4.27%  github.com/prometheus/prometheus/tsdb/index.(*MemPostings).Delete.func1
    1.90GB  3.79% 53.75%     1.90GB  3.79%  github.com/prometheus/prometheus/tsdb.NewCircularExemplarStorage
    1.83GB  3.67% 57.42%     4.41GB  8.82%  github.com/prometheus/prometheus/model/labels.(*ScratchBuilder).Labels
    1.80GB  3.60% 61.02%     1.80GB  3.60%  github.com/prometheus/prometheus/scrape.(*scrapeCache).addDropped
```

**Comment by bboreham** [comment]
Hi, I hope to give a lightning update at PromCon today - please post any statistics from people trying it out.

**Comment by prymitive** [comment]
> Hi, I hope to give a lightning update at PromCon today - please post any statistics from people trying it out.

I'm on leave most of September so I didn't pay much attention to these instances but here's a screenshot of a comparison for our two instances - one vanilla and the other is deduplabels.
It covers last 30 days and you can clearly see the moment when one instance was converted to deduplabels (410dm7). Up to that point both had pretty much same usage patterns, after than deduplabels uses less memory but ~10% more cpu. Memory usage of deduplabels is a lot less flat - which isn't unexpected given the need to obsessionally swap symbols tables etc.

screencapture-grafana-cfdata-org-d-em0ga5f4k-prometheus-performance-comparison-2024-09-12-11_29_13 (1)

**Comment by prymitive** [comment]
Here's a latest screenshot. Although there is a reduction in memory usage dedupelabels is a lot more unpredictable with how much memory will be used, compared to pretty much flat memory usage of stringlabels, which is not unexpected given how dedupelabels works.
I don't think I would be very eager to swap stringlabels to dedupelabels on our servers because of that, some memory savings are great and we can spare extra ~10% cpu to gain it, but I much more value stable resource usage pattern of stringlabels over that. Plus there would always be a worry that some service will hit the worst case scenario usage pattern for dedupelabels causing occasional OOM kills.

screencapture-grafana-cfdata-org-d-em0ga5f4k-prometheus-performance-comparison-2024-10-02-10_42_35

**Comment by bboreham** [comment]
If I understand your charts correctly, there is a point around 09/16 where `dedupelabels` memory goes ~5GB above `stringlabels`.  This has to be a bug somewhere; nothing about the design would lead to that result.

A profile at the time where memory goes higher would be useful.
